source/http_message.c

```diff
@@ -0,0 +1,238 @@
+ */
+struct http_message_binding {
+    struct aws_http_message *native;
+
+    /* Weak reference proxy to python self. */
+    PyObject *self_proxy;
+
+    /* Native input stream to read from HttpMessage._body_stream */
+    struct aws_input_stream *native_body_stream;
+};
+
+static void s_http_message_capsule_destructor(PyObject *capsule) {
+    struct http_message_binding *message = PyCapsule_GetPointer(capsule, s_capsule_name_http_message);
+
+    /* Note that destructor may be cleaning up a message that failed part-way through initialization */
+    aws_http_message_destroy(message->native);
+    aws_input_stream_destroy(message->native_body_stream);
+    Py_XDECREF(message->self_proxy);
+
+    aws_mem_release(aws_py_get_allocator(), message);
+}
+
+/**
+ * Update the aws_http_message based on its HttpRequest/HttpResponse Python self.
+ * Returns false and sets python error if failure occurred.
+ * Note that if failure occurred the aws_http_message might be in a
+ * "partially copied" state and should not be used.
+ */
+static bool s_http_message_update_from_py(struct http_message_binding *message) {
+
+    bool success = false;
+
+    /* Objects that need to be cleaned up whether or not we succeed */
+    PyObject *message_self = NULL;
+    PyObject *headers = NULL;
+    PyObject *map = NULL;
+    PyObject *method = NULL;
+    PyObject *path = NULL;
+
+    /* Doing a lot of queries so grab a hard reference */
+    message_self = PyWeakref_GetObject(message->self_proxy);
+    if (!message_self) {
+        goto done;
+    }
+    Py_INCREF(message_self);
+
+    headers = PyObject_GetAttrString(message_self, "headers");
+    if (!headers) {
+        goto done;
+    }
+
+    map = PyObject_GetAttrString(headers, "_map");
+    if (!map) {
+        goto done;
+    }
+
+    /* Clear existing headers, before adding the new values. */
+    size_t num_headers = aws_http_message_get_header_count(message->native);
+    for (size_t i = 0; i < num_headers; ++i) {
+        aws_http_message_erase_header(message->native, num_headers - (1 + i)); /* erase from back to front */
+    }
+
+    /* Copy in new header values */
+    struct aws_http_header header;
+    Py_ssize_t map_pos = 0;
+    PyObject *header_name = NULL;
+    PyObject *values_list = NULL;
+    while (PyDict_Next(map, &map_pos, &header_name, &values_list)) {
+        header.name = aws_byte_cursor_from_pystring(header_name);
+        if (!header.name.ptr) {
+            PyErr_SetString(PyExc_TypeError, "Header name is invalid");
+            goto done;
+        }
+
+        if (!PyList_Check(values_list)) {
+            PyErr_SetString(PyExc_TypeError, "Header values should be in a list");
+        }
+
+        Py_ssize_t num_values = PyList_GET_SIZE(values_list);
+        for (Py_ssize_t value_i = 0; value_i < num_values; ++value_i) {
+            PyObject *name_value_pair = PyList_GET_ITEM(values_list, value_i);
+            PyObject *header_value = PyTuple_GetItem(name_value_pair, 1);
```

Comment:
These 2 functions being named differently gave me cancer

Reply:
The first one is the can't-fail-macro version, because I've already validated the length of the list I'm iterating.

The second one CAN fail, I'm checking something that I'm assuming is a tuple with 2 items.

I'll add comments